### PR TITLE
feat: use Otel SDK 1.2/0.28

### LINKED ISSUE #983: contrib instrumentations are not compatible with SDK 1.2.0/0.28.0
### What version of OpenTelemetry are you using?

SDK 1.2.0/0.28.0
instrumentation-ioredis 0.28.0

### What version of Node are you using?

node.js 14 and 16 (both fail)

### What did you do?
Use latest SDK packages together with latest instrumentation-ioredis.

### What did you expect to see?

no compile error

### What did you see instead?

compile error `error TS2322: Type 'IORedisInstrumentation' is not assignable to type 'InstrumentationOption'.`

### Additional context

I think the reason is that `instrumentation-ioredis` depends on `"@opentelemetry/instrumentation": "^0.27.0"`. As far as I know `^` ranges for `0.x` versions update only in the patch version. As a result we get two versions of `@opentelemetry/instrumentation` in the node modules tree and typescript considers classes from different modules as different even if the shape is the same.

extract from package.json:
```
  "dependencies": {
    "@opentelemetry/api": "1.0.4",
    "@opentelemetry/instrumentation": "0.28.0",
    "@opentelemetry/instrumentation-ioredis": "0.28.0",
    "@opentelemetry/sdk-trace-base": "1.2.0"
  },
  "devDependencies": {
    "@types/node": "17.0.29",
    "typescript": "4.6.3"
  }
```` 
typescript reproducer:
```typescript
import { IORedisInstrumentation } from "@opentelemetry/instrumentation-ioredis";
import { registerInstrumentations } from "@opentelemetry/instrumentation";

registerInstrumentations({
  instrumentations: [
    new IORedisInstrumentation()
  ]
});
```

compile error:
``` 
index.ts:6:5 - error TS2322: Type 'IORedisInstrumentation' is not assignable to type 'InstrumentationOption'.
  Type 'IORedisInstrumentation' is not assignable to type 'Instrumentation'.
    Types of property 'setMeterProvider' are incompatible.
      Type '(meterProvider: import("C:/work/otel-compat/node_modules/@opentelemetry/instrumentation-ioredis/node_modules/@opentelemetry/api-metrics/build/src/types/MeterProvider").MeterProvider) => void' is not assignable to type '(meterProvider: import("C:/work/otel-compat/node_modules/@opentelemetry/api-metrics/build/src/types/MeterProvider").MeterProvider) => void'.
        Types of parameters 'meterProvider' and 'meterProvider' are incompatible.
          Type 'import("C:/work/otel-compat/node_modules/@opentelemetry/api-metrics/build/src/types/MeterProvider").MeterProvider' is not assignable to type 'import("C:/work/otel-compat/node_modules/@opentelemetry/instrumentation-ioredis/node_modules/@opentelemetry/api-metrics/build/src/types/MeterProvider").MeterProvider'.
            The types of 'getMeter(...).createObservableGauge' are incompatible between these types.
              Type '(name: string, callback: ObservableCallback, options?: MetricOptions | undefined) => void' is not assignable to type '(name: string, options?: MetricOptions | undefined, callback?: ((observableResult: ObservableResult) => void) | undefined) => ObservableBase'.
                Types of parameters 'callback' and 'options' are incompatible.
                  Type 'MetricOptions | undefined' is not assignable to type 'ObservableCallback'.
                    Type 'undefined' is not assignable to type 'ObservableCallback'.

6     new IORedisInstrumentation()
```

### PR DESCRIPTION
**Pull request by Flarna**
## Which problem is this PR solving?

fixes #983

## Short description of the changes

use ^0.28.0 in dependencies to SDK experimental packages
use 1.2.0 in dev-dependencies

## Checklist

- [ ] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.

### CODE REVIEW

**Comment by Flarna** [comment]
I tried to keep the changes low here. I have not adapted the host-metrics package as I'm not that familiar with metrics API.

There are still some dependencies to SDK packages like `"@opentelemetry/core": "^1.0.1"` which work but I'm not sure if we should keep them that way. Is see two options:
* stay at minimum version and allow newer (e.g. `^1.0.0`) like we do it with API
* depend on at latest SDK or even pin the minor version

**Comment by blumamir** [comment]
> * stay at minimum version and allow newer (e.g. `^1.0.0`) like we do it with API

I prefer this as it will allow users to update the SDK (patch / minor) without updating each instrumentation they use.

> * depend on at latest SDK or even pin the minor version

I think that if an instrumentation does not use any feature from a new minor version, it is cleaner to allow it to use old SDK so it's less hassel for users to update SDK version.

File: detectors/node/opentelemetry-resource-detector-aws/package.json [hunk 1]
```diff
@@ -42,7 +42,7 @@
   },
   "devDependencies": {
     "@opentelemetry/api": "1.0.2",
-    "@opentelemetry/contrib-test-utils": "^0.29.0",
+    "@opentelemetry/contrib-test-utils": "0.29.0",
```
[the remark this answers is not in the document]
Comment (Flarna) [reply detectors/node/opentelemetry-resource-detector-aws/package.json:45]: Forgot to save `detectors/node/opentelemetry-resource-detector-alibaba-cloud/package.json`. It's corrected now. But I haven't found other places.

File: detectors/node/opentelemetry-resource-detector-aws/package.json [hunk 2]
```diff
@@ -42,7 +42,7 @@
   },
   "devDependencies": {
     "@opentelemetry/api": "1.0.2",
```
[the remark this answers is not in the document]
Comment (Flarna) [reply detectors/node/opentelemetry-resource-detector-aws/package.json:44]: I considered to update also the API dev dependency but as it is not needed to get contrib working I skipped it.
It was already inconsistent before. I guess automation stopped working at some time and this is the result.

I vote for a cleanup in a separate PR.

I'm not sure what is better min or latest version. On the one hand we write ^1.0.0 in peer deps to allow users to use it but we never verify that no 1.1 api/type is used anywhere.

File: detectors/node/opentelemetry-resource-detector-docker/package.json [hunk 3]
```diff
@@ -36,7 +36,7 @@
   },
   "devDependencies": {
```
[the remark this answers is not in the document]
Comment (Flarna) [reply detectors/node/opentelemetry-resource-detector-docker/package.json:37]: corrected it here and hopefully at a all other places I missed

**Comment by Flarna** [comment]
In general I'm fine with allowing older SDKs. Main issue is that there is no test actually verifying that it works as we install newest in our tests so it's easy to miss.

**Comment by dyladan** [comment]
> In general I'm fine with allowing older SDKs. Main issue is that there is no test actually verifying that it works as we install newest in our tests so it's easy to miss.

We can use test all versions but the CI will take a very long time. We should do it at least for releases though

**Review by blumamir — APPROVED** [review]

**Review by vmarchaud — APPROVED** [review]

**Comment by legendecas** [comment]
I've updated the host-metrics at https://github.com/open-telemetry/opentelemetry-js-contrib/pull/990.

**Review by legendecas — APPROVED** [review]

**Comment by vmarchaud** [comment]
I think if @dyladan is fine with it we can merge

**Review by Rauno56 — APPROVED** [review]
That's awesome, thanks!

**Pull request merged**